add a release checklist

gasche opened the pull request:
Currently this list isn't publicly available, it sits in various
different versions on @damiendoligez's filesystems. He sent me a copy
when I took care of some of the recent releases. The present
presentation is a result of significant cleanups and changes to the
checklist -- in particular, some mistakes may have jumped in.

This is not a scripted process, it is very informal and it is likely
that there are some mistakes/omissions in the list. Yet, it sounds
better to have it somewhere in the source repository than not have it
around at all. It was certainly helpful to me, and it probably would
be to other release-help volunteers.

(The list is put in tools/ because there isn't a clearly better place
for it. This choice was suggested by Damien.)

gasche committed:
add a release checklist

Currently this list isn't publicly available, it sits in various
different versions on @damiendoligez's filesystems. He sent me a copy
when I took care of some of the recent releases. The present
presentation is a result of significant cleanups and changes to the
checklist -- in particular, some mistakes may have jumped in.

This is not a scripted process, it is very informal and it is likely
that there are some mistakes/omissions in the list. Yet, it sounds
better to have it somewhere in the source repository than not have it
around at all. It was certainly helpful to me, and it probably would
be to other release-help volunteers.

(The list is put in tools/ because there isn't a clearly better place
for it. This choice was suggested by Damien.)

shindere commented:
Many thanks @æasche, I find this a great idea to have the list in the
repository.

Wouldn't it make sense to create a doc directory where we could put
stuff like HAKING, BOOSTRAPING etc.,

shindere commented:
"Warn Gabriel Scherer (to make a pass on Changes) and OCamlLabs (for OPAM
testing).": I would replace the names by `caml-devel@içnria.fr` which looks
more stable to me.

The file skips from step "1: check repository state" to step
"3: build, refresh dependencies, sanity
checks": step 2 is missing.

"check that .depend files have no absolute path in them": well, if one day
we have a global .depend there will be slashes in it, although paths are not
absolute.

./tools/check-symbol-names asmrun/*.a byterun/*.a

should now be:

./tools/check-symbol-names runtime/*.a

"# must have empty output and result 0": I'd replace result by return

"at this point, VERSION contains N+devD": I think here you are referring to
the VERSION file at the root of the repository but it is ambiguous because
you have introduced a VERSION environemnt variable earlier so I'd say "the
VERSION file (at the root of the repository)" or something clearer.

About the OPAM part: for the non-expert I am, I don't understand the
instructions. In which directories should the mentionned commands be run?

Why do the compression commands (gzip and xz) use time?

"## 8: upload the archive and compute checksums": archives rather than
archive?

About OS X packaging: the `package-macosx` target is no longer available.

pmetzger commented on tools/release-checklist:
```diff
@@ -0,0 +1,327 @@
+For production releases, you should get in touch with ocaml.org to
+organize the webpage for the new release. See
+
+  <https://github.com/ocaml/ocaml.org/issues/819>
```
We need a bit more on the list than this, I think, because for 4.06 the web pages were not fixed for quite some time and not everything was noticed at once.

gasche replied:
In the issue someone volunteered to write more detailed steps, and I'm waiting for them to include here. ;-)

pmetzger commented on tools/release-checklist:
````diff
@@ -0,0 +1,327 @@
+Happy hacking,
+
+-- Damien Doligez for the OCaml team.
+```
````
Missing a trailing carriage return.

pmetzger commented:
This is a very good list! Thank you!

For MacOS, it would be good to also inform the people who manage the Brew formula for OCaml (I don't know who that is) and to inform the MacPorts maintainer (which is me.)

xavierleroy commented:
I'm slightly uncomfortable at the quantity of information about our server infrastructure that is made public here.  But I guess the cat is out of the bag already...

pmetzger commented:
> I'm slightly uncomfortable at the quantity of information about our server infrastructure that is made public here.

It could be removed from the file still. Not that many people have seen it.

gasche commented:
I removed details from the file. I'm not fully satisfied without a reasonable place to store the remaining bits (right now they are on my machine and Damien's, we should think of something more portable), but that's a good compromise I guess.

avsm commented:
As we open up the opam hardware infrastructure to more developers (especially non-x86), I wanted a place to store server details in a slightly less public place. I could create a private repository along the lines of https://github.com/ocaml/infrastructure-private (to go with the public https://github.com/ocaml/infrastructure)

shindere commented:
Before I forget: hadn't we decided to change the executable magic number for
each production release?

Shouldn't that be mentionned / explained in the list?

dra27 commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+# for actual releases: check and change the Changes header
+#  (remove "next version" and add a date)
+git add VERSION
+git commit -m "last commit before tagging $VERSION" 
```
Whitespace at end of line

dra27 commented on tools/release-checklist:
````diff
@@ -0,0 +1,326 @@
+
+```
+ssh <<archive machine>> "mkdir -p $DIST/notes"
+scp INSTALL.adoc LICENSE README.adoc README.win32.adoc Changes <<archive machine>>:$DIST/notes/
````
This line is over-long (though perhaps that just means `ocaml-typo=long-line` is wanted in `.gitattributes` for this file)

dra27 commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+Happy hacking,
+
+-- Damien Doligez for the OCaml team.
+   
```
Stray whitespace here

dra27 reviewed (dismissed):
Since #1148 will be merged soon, I've done a scan of non-rebase-requiring PRs for check-typo so that we don't merge things which then break Travis/Inria's trunk run.

This file definitely wants `ocaml-typo=missing-header` in `.gitattributes` and probably `long-line` as well.

shindere commented:
David Allsopp (2018/06/30 09:48 -0700):
> dra27 requested changes on this pull request.
> 
> Since #1148 will be merged soon, I've done a scan of non-rebase-requiring PRs for check-typo so that we don't merge things which then break Travis/Inria's trunk run.
> 
> This file definitely wants `ocaml-typo=missing-header` in
> `.gitattributes` and probably `long-line` as well.

Coudln't we do without ``long-line` ? Or is it because of URLs?

dra27 commented:
@shindere - it's line 188, so it could be altered. URLs already can't cause long-line to trigger (IIRC the rule is that if the line is only over-long because of one URL then long-line is suppressed - i.e. you should still be trying to keep to 80 characters if possible!)

gasche commented:
Thanks for all the feedback. I will do another turn of the crank at some point. I plan to only fix the minor issues, though, not add new content to the file. It will remain imperfect, and we can improve it gradually.

I also need to get confirmation by @damiendoligez that the reordering of steps (compared to the initial spaghetti file) is valid, but I may have to go get it in person. It's hard to spend time on release files outside releases, and we have other things to do during the release.

shindere commented:
@gasche: Damien confirmed that the package-macosx thing is no longer
relevant, as I suggested the other day.

shindere commented:
David Allsopp (2018/07/02 02:07 -0700):
> @shindere - it's line 188, so it could be altered. URLs already can't
> cause long-line to trigger (IIRC the rule is that if the line is only
> over-long because of one URL then long-line is suppressed - i.e. you
> should still be trying to keep to 80 characters if possible!)

@dra27 thanks!

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+# at this point, VERSION contains N+devD
+# increment VERSION into N+dev(D+1); for example,
+#   4.07.0+dev8-2018-06-19 => 4.07.0+dev9-2018-06-26
+# for actual releases: check and change the Changes header
```
"actual releases" should be "production releases"

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+git commit -m "change VERSION for $VERSION" -a
+git tag -m "release $VERSION" $VERSION
+
+# for actual releases, change VERSION into (N+1)+dev0; for example,
```
Again, this should be "production releases".

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+
+# for actual releases, change VERSION into (N+1)+dev0; for example,
+#   4.08.0 => 4.08.1+dev0
+# for release candidates, use N+dev(D+1) instead; for example,
```
And this should be "testing releases".

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+
+## 13: update Mantis
+
+Update Mantis by adding $MAJOR.$MINR.$BUGFIX as a version number for reports.
```
$MINR is missing an O

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+
+Update Mantis by adding $MAJOR.$MINR.$BUGFIX as a version number for reports.
+
+## 14: build and release OSX binaries
```
As @shindere said, this whole section is dead code.

gasche replied:
Well there goes my mild guilt at ignoring it each time...

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,326 @@
+
+See the email announce templates at the end of this file.
+
+## 13: update Mantis
```
You should probably swap steps 12 and 13.

damiendoligez reviewed (commented):
A few small things. The order of steps looks OK.

gasche commented:
Thanks! Speaking of "update Mantis", I have never done it myself, I don't think I have the necessary rights, and I don't know how to do it; if you have an URL at hand to cut the link chase, I could include it in the notes.

damiendoligez commented:
You would go to https://caml.inria.fr/mantis/manage_proj_edit_page.php?project_id=1 , log in, then scroll down to "Versions".

gasche commented:
Thanks! I can confirm that I don't have the privilege level to do it.

gasche commented:
Thanks to everyone's careful feedback, I did another pass on the file, and I believe that all concerns voiced so far have been addressed.

In particular, I followed @shindere's excellent suggestion to include a remark on bumping magic numbers (I was glad to add a section 2 instead of having to manually renumber). Instead of giving too much details in that file, I created a utils/HACKING.adoc file with a description of the magic number bumping policy.

@dra27: I was perplexed for a while by the fact that `./tools/check-typo utils/HACKING.adoc` would return a missing-header error despite `*.adoc` being covered in `.gitattributes`. Then I noticed that I had forgotten to `git add` the file; my understanding is that it is then not taken into account by whatever gitattributes-querying mechanism that you are using. Could you maybe have `check-typo` emit a warning when interrogated on a file that is not part of the git index? (It doesn't need to be a scary warning, as it is a reasonable use-case during development of a new file, but still it would have helped me.)

dra27 commented:
@gasche - indeed, the problem is that `git check-attr` doesn’t apply to files not in git, or at least registered in the index (I think if you had done `git add -N` it would also have been fine).

I’ll see if there’s a way to get hypothetical Git attributes for non-added files, but otherwise your idea of a warning is good.

pmetzger commented:
I think it would be wise to have a step, about a week before release, notifying maintainers for various packaging organizations (like Debian or Arch or Brew or what have you) that they should check that the last release candidate works okay on their platform in preparation for release. It might be a bit late in the 4.07 process for that, but it could be done for 4.08.

shindere commented:
Gabriel Scherer (2018/07/03 13:54 +0000):
> In particular, I followed @shindere's excellent suggestion to include
> a remark on bumping magic numbers (I was glad to add a section 2
> instead of having to manually renumber). Instead of giving too much
> details in that file, I created a utils/HACKING.adoc file with a
> description of the magic number bumping policy.


Very well and cleverly done, @gasche! And I'm not saying this because
the suggestion was mine. ;-)

xavierleroy commented:
> I created a utils/HACKING.adoc file with a description of the magic number bumping policy.

`utils/` is a source directory.  I'm not too happy about putting documentation there.

pmetzger commented:
> `utils/` is a source directory. I'm not too happy about putting documentation there.

I think `HACKING` style files are a bit different, though. They are not information about using the code, but rather guides to the code itself.

gasche commented:
@pmetzger: my expectation is that distribution maintainers would follow caml-list or caml-announce, and thus be notified of release candidates which generally happen about one week before the release. Do we really need a separate communication step?

@xavierleroy: we already have HACKING.adoc files in `parsing/` and `typing/`. I generally like having developer-oriented documentation close to the thing they are documenting.

shindere commented:
Gabriel Scherer (2018/07/03 09:29 -0700):
> @xavierleroy: we already have HACKING.adoc files in `parsing/` and
> `typing/`. I generally like having developer-oriented documentation
> close to the thing they are documenting.

Well then the magic number stuff could as well be in `runtime/` or even
` runtime/caml/` since it's a runtime thing after all.

I am not saying it should be there, my point is trying to show that it
may not always make good sense to keep the doc close to the source,
especially when something is defined in several places. And to (perhaps)
make this argument stronger: with autoconf the magic number could (I
think it should, actually) be defined in `configure.ac`, so the files
that need it would be generated and then the document file would have to
be moved I don't really know where.

Wouldn't it be better to have a `doc/` subdirectory where all the
developer doc could go? The advantage would be that it would be gathered
and that the sources could evolve on their won.

pmetzger commented:
@gasche:

> my expectation is that distribution maintainers would follow caml-list or caml-announce, and thus be notified of release candidates which generally happen about one week before the release. Do we really need a separate communication step?

Yes, you're probably right that it is unnecessary. That said, I would also announce release candidates on discuss.ocaml.org, just in case.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+export VERSION=$MAJOR.$MINOR.$BUGFIX$PLUSEXT
+
+export REPO=http://github.com/ocaml/ocaml
+export DIST=<<dist path>>/ocaml/ocaml-$MAJOR.$MINOR
```
Why invent a new notation for metavariables where a plain variable does the job better? If you just write `$DIST_PATH` instead of `<<dist path>>` this file will be easier to use by the release manager.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+  # ocamldoc/stdlib_non_prefixed depends on 'world'
+
+# check that .depend files have no absolute path in them
+find . -name .depend | xargs grep ' /'
```
Please add a comment `# must have empty output`.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+#   4.07.0+dev8-2018-06-19 => 4.07.0+dev9-2018-06-26
+# for production releases: check and change the Changes header
+#  (remove "next version" and add a date)
+git add VERSION
```
Should be `git add VERSION Changes` since the `Changes` file may be changed at this point.

damiendoligez commented on tools/release-checklist:
````diff
@@ -0,0 +1,352 @@
+
+```
+# copy foo+rc1+... switches into foo+rc2+...
+for f in *+rc1*; do mkdir -p $f $(echo $f | sed 's/+rc1/+rc2/g'); done
````
I don't think you need the `$f` argument to mkdir.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+# copy foo+rc1+... switches into foo+rc2+...
+for f in *+rc1*; do mkdir -p $f $(echo $f | sed 's/+rc1/+rc2/g'); done
+for f in *+rc1*/*; do cp $f $(echo $f | sed 's/+rc1/+rc2/g'); done
+git add *+rc2* # import into git (the files still use +rc1 all over)
```
This line doesn't work for a production release.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+for f in *+rc1*/*; do cp $f $(echo $f | sed 's/+rc1/+rc2/g'); done
+git add *+rc2* # import into git (the files still use +rc1 all over)
+
+sed -i "s/+rc1/+rc2/g" *rc2*/*
```
Doesn't work on MacOS because sed wants an argument after `-i`.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+opam1 repository organization, an opam2 repository will have
+a different layout.
+
+From a clone of the opam-repository, in `compilers/$MAJOR.$MINOR.$BUGFIX`:
```
Not just a clone: you need a branch in order to make a PR against opam-repository.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+git diff # inspect the result
+git add *+rc2*
+
+sed -i "s/First release candidate/Second release candidate/g" *+rc2*/*
```
Same remark about `sed -i`.

damiendoligez commented on tools/release-checklist:
````diff
@@ -0,0 +1,352 @@
+
+git commit -m "OPAM switches for $VERSION"
+```
+
````
to add: push the branch to GitHub and make a PR

damiendoligez commented on tools/release-checklist:
````diff
@@ -0,0 +1,352 @@
+## 8: upload the archives and compute checksums
+
+```
+scp ocaml-$VERSION.tar.{xz,gz} <<archive machine>>:$DIST
````
Again, replace the metavariable with a real variable. There are more in the file, listing them all would make too much noise so I'll stop here.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+# check new-md5s to ensure that they look right, and then
+cat new-md5s >> MD5SUM
+# if everything worked well,
+rm MD5SUM.old
```
You should remove `new-md5s` as well.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+cp SHA512SUM SHA512SUM.old
+sha512sum ocaml-$VERSION.tar.{xz,gz} > new-sha512s
+cat new-sha512s >> SHA512SUM
+rm SHA512SUM.old
```
And remove `new-sha512s` too.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+
+# upload manual checksums
+ssh <<archive machine>> "cd $DIST; md5sum ocaml-$BRANCH-refman* >>MD5SUM"
+ssh <<archive machine>> "cd $DIST; sha256sum ocaml-$BRANCH-refman* >>SHA512SUM"
```
Should be `sha512sum`.

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+mkdir -p ocaml-$BRANCH
+cd ocaml-$BRANCH
+wget http://caml.inria.fr/pub/distrib/ocaml-$BRANCH/ocaml-$BRANCH-refman-html.tar.gz
+tar -xzvf manual-ocaml-$BRANCH-refman-html.tar.gz # this extracts into htmlman/
```
remove the extraneous `manual-`

damiendoligez commented on tools/release-checklist:
```diff
@@ -0,0 +1,352 @@
+mv htmlman/* -t . # move HTML content to docs/manual-caml-$BRANCH
+rmdir htmlman
+
+cd <<www>>/caml/pub/docs/manual-ocaml
```
Remove the last part (`manual-ocaml`).

damiendoligez commented on utils/HACKING.adoc:
```diff
@@ -0,0 +1,52 @@
+
+We say that we "bump" a magic number when we update its version part
+in config.mlp. To bump all magic numbers is to increment the version
+of all kinds of magic numbers.
```
You should write "every kind", as "all kinds of" has another meaning in English.

damiendoligez commented on utils/HACKING.adoc:
```diff
@@ -0,0 +1,52 @@
+
+This should preferably be done just before the first testing release
+(the first beta, or the first rc if there is no beta) of the new major
+release. We want to to happen after all format-breaking changes have
```
"we want to to" -> "we want it to"

damiendoligez reviewed (commented):
A few suggestions from testing the checklist on 4.07.0.

gasche commented:
@damiendoligez thanks, I made another pass of changes.

(The opam scripts still don't work in the edge cases of creating a new directory or the `+trunk` stuff, but this may have to suffice for now.)

dra27 commented:
Some notes for opam 2.0.0 packages:
 - The next trunk version of the `ocaml` meta package has to be created **when the new release is branched**. So, for example, when `4.07` was branched, the `ocaml.4.08.0` package became necessary (see https://github.com/ocaml/opam-repository/commit/fa335884535948456330b3f9923c489281821219).
 - The changes presently made to `compilers` in OPAM 1.2.2 get made to the `ocaml-base-compiler` package. This is very easy at release, as the beta and rc definitions will already have been in `ocaml-variants`, so they just copied and the version number updated.
 - Since homebrew will upgrade about 3 seconds after the release, it's important to create an `ocaml-system` package for the new release (see https://github.com/ocaml/opam-repository/pull/12357)

pmetzger commented:
A suggestion: in the few weeks before a release, some effort should be made to make sure that the most commonly installed third party packages (say things like menhir and utop and lwt and the like) have opam packages compatible with the new release. This shouldn't be a blocking step, but it would be good to make a best-efforts attempt.

gasche commented:
I already gave an opinion on this topic on [the relevant Discuss thread](https://discuss.ocaml.org/t/ocamls-release-process/2297/5): I think this is outside the responsibility of the core compiler team, and in any case while that may be part of the decision in the *timing* of a release, it needs not be part of the *compiler release process*.

gasche commented:
I just rebased the PR on trunk and fixed all pending comments.

(@dra27, I didn't include the opam2 instructions, sorry. Hopefully this can come as a separate PR after this one is merged.)

gasche commented:
(No one has seemed interested in deciding to approve it, yet that would be necessary for it to be merged.)

shindere commented:
Gabriel Scherer (2018/08/19 12:47 -0700):
> (No one has seemed interested in deciding to approve it, yet that
> would be necessary for it to be merged.)

I'll go through it tomorrow @gasche and will gladly approve it if
everything seems fine to me.

dbuenzli commented:
Not a prerequisite for merge but if the [`ocaml-manual`](https://opam.ocaml.org/packages/ocaml-manual/) opam package could be updated in the process that would be nice. 

(Even better would be to install the manual, `Changes` and `README` file in the `opam var ocaml:doc` directory when the `ocaml` package is installed. This allows users to simply `odig changes ocaml` to get the release notes of their current opam compiler version without having to track them down on the interweb).

shindere commented:
The Changes entry has the wrong GPR number. It says GPR#1877 but should
say GPR#1866.

If I may nitpick: I'd use an uppercase A for the first letter of the
commit message. It's like a title after all.

Currently utils/HACKING.adoc says that magic numbers are defined in
config.mpl. It's true but the one for executables is also defined in
runtime/caml/exec.h. Shouldn't that be documented?

And in this light, why adding one mire HACKING file rather than
modifying the one at the top level?

"This lets the compiler differenciates": differentiate?

"However, it is more dangerous to forget to update a number than to
update too often": I'd say to update *it*

"If we update too often": if updated too often, to avoid the repetition?

"If we don't update +often enough": it, again?

Well, the wording of all this could probably be improved, e.g. "too
frequent/infrequent" updates...

Coming to the release checklist itself, I think I already did this
remark so I'm gonna be repeating myself. To me, it's a bad idea to write
names of persons in such places, I'm referring the "Warn Gabriel Scherer
(to make a pass on Changes) and OCamlLabs (for OPAM testing)."
sentences. I'd rather recommend to announce this on the develpment list,
because this looks more stable to me than names of persons or
institutions that may have to be updated more frequently than the name
of a mailing-list. It's just too easy to forget such updates. And even
if they are not forgotten, why create the need for such a task when it
could be avoided at the first place?

"git status  # check that the repo is in a clean state": I'd be more
precise here and add *local*, i.e. "check that the local repo is in a
clean state".

"See the HACKING file of `utils/` for documentation on how to bump the
magic numbers.": ok so the other file definitely needs to mention
exec.h, I think.

"make distclean": I'd add a call to git clen, too, because make
distclean may forget files. And maybe if that happens it means that make
distclean should be fixed.

About section 5, I am wondering whether it wouldn't be clearer to split
its content into two or three subsections: Beta, Rc, Production. I foudn
this section not so easy to understand.

"I use a semi-automated process": donn't we want to get rid of the first
person in the documentation?

How about "A semi-automated process can be used", for instance?

Perhaps you could add a cd command at the beginning of the opam part to
make it clearer that you work on the opam repository? And then another
cd at the beginning of section 7 to make clear that one has to get back
to the ocaml repository?

"To update the checksum files on the remote host, we first upload the
release environment as a script.": well in that case, perhaps you could
write the environment in a script from the beginning, in section 0, and
then source it. Then here you'd just have to copy it. I think that could
also me convenient when switching from the Beta or Rc to the production
version, because you'd just have to update that script, which could even
become part of the repo, or atleast a template could be under version
control.

that way you would avoid the check step that you have just after the cat
commands.

thanks for your great work and commitment on this!

damiendoligez committed:
checklist fixes

damiendoligez committed:
take @shindere's remarks into account

shindere commented:
Thanks, @damiendoligez

damiendoligez commented:
Added a few fixes after trying it out on 4.07.1+rc1, made changes to take @shindere's remarks into account, and rebased.
I'll merge after CI gives the green light.
